hermes/src/network/rpc/ws.rs

```diff
@@ -0,0 +1,290 @@
+    super::types::{
+        PriceIdInput,
+        RpcPriceFeed,
+    },
+    crate::store::Store,
+    anyhow::Result,
+    axum::{
+        extract::{
+            ws::{
+                Message,
+                WebSocket,
+                WebSocketUpgrade,
+            },
+            State,
+        },
+        response::IntoResponse,
+    },
+    dashmap::DashMap,
+    futures::{
+        stream::{
+            SplitSink,
+            SplitStream,
+        },
+        SinkExt,
+        StreamExt,
+    },
+    pyth_sdk::PriceIdentifier,
+    serde::{
+        Deserialize,
+        Serialize,
+    },
+    std::{
+        collections::HashMap,
+        sync::atomic::{
+            AtomicUsize,
+            Ordering,
+        },
+    },
+    tokio::sync::mpsc,
+};
+
+
+pub async fn ws_route_handler(
+    ws: WebSocketUpgrade,
+    State(state): State<super::State>,
+) -> impl IntoResponse {
+    ws.on_upgrade(|socket| websocket_handler(socket, state))
+}
+
+async fn websocket_handler(stream: WebSocket, state: super::State) {
+    let ws_state = state.ws.clone();
+    let id = ws_state.subscriber_counter.fetch_add(1, Ordering::SeqCst);
+
+    let (sender, receiver) = stream.split();
+
+    // TODO: Use a configured value for the buffer size or make it const static
+    // TODO: Use redis stream to source the updates instead of a channel
+    let (tx, rx) = mpsc::channel::<Vec<PriceIdentifier>>(1000);
+
+    ws_state.subscribers.insert(id, tx);
+
+    log::debug!("New websocket connection, assigning id: {}", id);
+
+    let mut subscriber = Subscriber::new(id, state.store.clone(), rx, receiver, sender);
+
+    subscriber.run().await;
+}
+
+pub type SubscriberId = usize;
+
+/// Subscriber is an actor that handles a single websocket connection.
+/// It listens to the store for updates and sends them to the client.
+pub struct Subscriber {
+    id:                      SubscriberId,
+    closed:                  bool,
+    store:                   Store,
+    update_rx:               mpsc::Receiver<Vec<PriceIdentifier>>,
+    receiver:                SplitStream<WebSocket>,
+    sender:                  SplitSink<WebSocket, Message>,
+    price_feeds_with_config: HashMap<PriceIdentifier, PriceFeedClientConfig>,
+}
+
+impl Subscriber {
+    pub fn new(
+        id: SubscriberId,
+        store: Store,
+        update_rx: mpsc::Receiver<Vec<PriceIdentifier>>,
+        receiver: SplitStream<WebSocket>,
+        sender: SplitSink<WebSocket, Message>,
+    ) -> Self {
+        Self {
+            id,
+            closed: false,
+            store,
+            update_rx,
+            receiver,
+            sender,
+            price_feeds_with_config: HashMap::new(),
+        }
+    }
+
+    pub async fn run(&mut self) {
+        loop {
+            if self.closed {
+                break;
+            }
+            if let Err(e) = self.handle_next().await {
+                log::error!("Subscriber {}: Error handling next message: {}", self.id, e);
```

Comment:
do you really want to kill the subscriber here ? seems like it might be simpler to have one way to terminate the connection (setting self.closed) and then move all error logging here. I note you have some code below that transforms errs into Oks which could go away if you logged here instead.

Reply:
The reason I did that was that close is not an error and the log there is on debug level and this one has error log level. Usually this should not happen and we need a higher log level for it.